### PR TITLE
Allow for extension version to specified in friendly iframe embeds

### PR DESCRIPTION
**Pull request by alabiaga**
fixes 1/2 tasks in #12429

support 'extensions' on client side for friendlyIframeSpec, allowing for extensions and their versions.

related: #7217

### CODE REVIEW

**Comment by alabiaga** [comment]
@lannka  @keithwrightbos @choumx friendly ping.

File: extensions/amp-a4a/0.1/test/test-amp-ad-utils.js [hunk 1]
```diff
@@ -18,6 +18,9 @@ import {data} from './testdata/valid_css_at_rules_amp.reserialized';
 import {getAmpAdMetadata} from '../amp-ad-utils';
 
 describe('getAmpAdMetadata', () => {
+  beforeEach(() => {
```
Comment (lannka) [inline extensions/amp-a4a/0.1/test/test-amp-ad-utils.js]: is this change needed?

**Comment by alabiaga** [comment]
@lannka Thanks for the review. As far as your comment regarding the metadata for ads, customElementExtensions is definitely used, I saw documentation under amp-cache-modifications that it is deprecated but still in use and that "extensions" is supported..though i don't see anything in the code that supports and reads that data from that field. Perhaps the intention was for it to replace the cusomtElementExtensions, and the modifications have just not been made. @keithwrightbos Let me know if this is the case and whether this would be the preferred way to support versions with extensions and I can make that change...or perhaps I am wrong and didn't read the code correctly. Thanks

**Comment by lannka** [comment]
@alabiaga that's what I see as well. Just want to confirm with @keithwrightbos because the server transformer needs to agree on the same format.
 
Meantime, please do have some manual test against `fake-ad.amp.html`, and play with extensions with different versions.

**Comment by alabiaga** [comment]
@lannka awaiting Keith's input on this and I've pinged him. I will do the manual testing in the meantime. I've left the tests broken until I get his feedback. Many thanks for your feedback!

**Comment by alabiaga** [comment]
confirmed that the extensions field in amp-ad-metadata is not used when determining which extensions to load but that the customElementExtensions is used. Thanks

**Comment by keithwrightbos** [comment]
Apologies for the delayed response.  We are currently using customElementExtensions (see https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/0.1/amp-a4a.js#L1601) which does not include the version, just the name of the extension.  However, we also return the extension in the format mentioned by @lannka and you should feel free to use it going forward.  Once fully deployed, we can remove sending the old customElementExtensions format.  Thanks

**Comment by alabiaga** [comment]
spoke to @keithwrightbos, the client code was never modified to support this. the ad stack does support the extensions field for amp-ad-metadata when the ad is served. I will add this support on the client side.

**Comment by jridgewell** [comment]
Oh, I think @newmuis actually did part of this?

**Comment by alabiaga** [comment]
@jridgewell Correct, you cc'ed me in the message thread when @newmuis was asking about this bug. He did address this but in the context of reloading extensions. This PR is addressing the versioning bug in the context of FIE and ads. Thanks for mentioning.

**Comment by alabiaga** [comment]
Friendly ping. this is ready for review.

File: src/service/extensions-impl.js [hunk 2]
```diff
@@ -445,13 +446,18 @@ export class Extensions {
     stubLegacyElements(childWin);
 
     const promises = [];
-    extensionIds.forEach(extensionId => {
+    extensionData.forEach(extension => {
+      let extensionId;
+      if (extension && typeof extension == 'object') {
```
Comment (jridgewell) [inline src/service/extensions-impl.js]: A4A should be passing down the `custom-element` fields directly. There's no reason for `extension-impl.js` to know anything about A4A objects.

File: src/service/extensions-impl.js [hunk 3]
```diff
@@ -631,6 +637,30 @@ export class Extensions {
...
 }
 
+/**
+ * @param {!../../extensions/amp-a4a/0.1/amp-ad-type-defs.CreativeMetaDataDef} metadata
+ * @param {string} extensionId
+ * @return {boolean}
+ */
+export function hasExtensionId(metadata, extensionId) {
+  let ext = [];
+  const {extensions, customElementExtensions} = metadata;
```
Comment (jridgewell) [inline src/service/extensions-impl.js]: Again, this file should not have to be aware of A4A objects.

File: src/friendly-iframe-embed.js [hunk 4]
```diff
@@ -629,3 +641,28 @@ export function whenContentIniLoad(elementOrAmpDoc, hostWin, rect) {
...
 }
+
+/**
+ * Preloads extensions. The extensions field is the preferred means as it
+ * supports versions.
+ * @param {?../extensions/amp-a4a/0.1/amp-ad-type-defs.CreativeMetaDataDef|!FriendlyIframeSpec} metaData
+ * @param {!Window} win
+ */
+export function preloadExtensions(metaData, win) {
+  if (metaData.extensions && metaData.extensions.length) {
```
Comment (jridgewell) [inline src/friendly-iframe-embed.js]: This "if one object type, do this, else if other type, do this" is icky. Why aren't we just standardizing a format to pass down?

Comment (alabiaga) [reply src/friendly-iframe-embed.js]: verified this is no longer used. Would've also helped if i read the comment here https://github.com/ampproject/amphtml/issues/7217#issuecomment-289822598, will remove customElementExtensions throughout codebase. Thanks

**Comment by alabiaga** [comment]
@jridgewell @keithwrightbos ping. Added changes to gate on experiment. Also added a user warn log call to check that the old field isn't used.

related: https://github.com/ampproject/amphtml/pull/18738

File: extensions/amp-a4a/0.1/template-validator.js [hunk 5]
```diff
@@ -101,4 +94,52 @@ export class TemplateValidator extends Validator {
...
+  addExtensionIfApplicable_(creativeMetadata, extensionId, extensionSrc) {
+    if (creativeMetadata) {
+      let ext = [];
+      const {extensions, customElementExtensions} = creativeMetadata;
+      if (extensions && extensions.length) {
+        ext = extensions;
+      } else if (customElementExtensions && customElementExtensions.length) {
+        ext = customElementExtensions;
+      }
+      if (ext.indexOf(extensionId) == -1) {
```
Comment (jridgewell) [inline extensions/amp-a4a/0.1/template-validator.js]: I don't believe this can work in the `extensions` case, they're objects not strings.

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 6]
```diff
@@ -463,13 +463,21 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
...
-          this.getAmpDoc(), this.iframe.contentWindow);
+    let ext = [];
+    if (creativeMetaData) {
+      const {extensions, customElementExtensions} = creativeMetaData;
+      if (extensions && extensions.length) {
+        ext = extensions;
+      } else if (customElementExtensions && customElementExtensions.length) {
+        ext = customElementExtensions;
+      }
+      if (ext.indexOf('amp-ad-exit') == -1) {
```
Comment (jridgewell) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: Ditto.

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 7]
```diff
@@ -913,13 +913,21 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
...
-          this.getAmpDoc(), this.iframe.contentWindow);
+    if (creativeMetaData) {
+      let ext = [];
+      const {extensions, customElementExtensions} = creativeMetaData;
+      if (extensions && extensions.length) {
+        ext = extensions;
+      } else if (customElementExtensions && customElementExtensions.length) {
+        ext = customElementExtensions;
+      }
+      if (ext.indexOf('amp-ad-exit') == -1) {
```
Comment (jridgewell) [inline extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: Ditto.

File: src/friendly-iframe-embed.js [hunk 8]
```diff
@@ -206,11 +220,24 @@ export function installFriendlyIframeEmbed(iframe, container, spec,
...
+    // TODO (alabiaga): remove customElementExtensions once extensions
+    // becomes preferred way.
+    let ext = [];
+    if (isExperimentOn(win, 'fie-metadata-extension') && spec.extensions) {
+      ext = spec.extensions;
+    } else if (spec.customElementExtensions) {
+      spec.customElementExtensions.forEach(extensionId => {
+        // src is not specified as legacy extensions will default to
+        // their 0.1 version. See extensions.installExtensionsInChildWindow.
+        ext.push({'custom-element': extensionId, 'src': null});
```
Comment (jridgewell) [inline src/friendly-iframe-embed.js]: `version` instead of `src`?

File: src/friendly-iframe-embed.js [hunk 9]
```diff
@@ -629,3 +656,32 @@ export function whenContentIniLoad(elementOrAmpDoc, hostWin, rect) {
...
+ * @param {!Window} win
+ */
+export function preloadExtensions(metaData, win) {
+  if (isExperimentOn(win, 'fie-metadata-extension')
+      && metaData.extensions && metaData.extensions.length) {
+    metaData.extensions.forEach(
+        extensionDef => {
+          const extensionId = extensionDef['custom-element'];
+          const version = extensionDef['src'];
+          Services.extensionsFor(win)
```
Comment (jridgewell) [inline src/friendly-iframe-embed.js]: Extract this part out of the loop (and if statements).

**Comment by lannka** [comment]
@alabiaga are you still on it? or shall we close?

**Comment by alabiaga** [comment]
@lannka unfortunately this CL has been stale as i haven't had time to fix the test. I'll try to resolve this week..otherwise i'll close it.

**Comment by mrjoro** [comment]
@alabiaga is this PR still active?

**Comment by alabiaga** [comment]
closing. these changes shouldn't be all too difficult to port over in another PR without the complications of merging.
